File: platform-sponge/src/main/java/taboolib/platform/SpongePlugin.java
package taboolib.platform;

import com.google.inject.Inject;
import org.jetbrains.annotations.Nullable;
import org.spongepowered.api.config.ConfigDir;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.game.state.*;
import org.spongepowered.api.plugin.PluginContainer;
import taboolib.common.LifeCycle;
import taboolib.common.TabooLibCommon;
import taboolib.common.io.IOKt;
import taboolib.common.platform.FunctionKt;
import taboolib.common.platform.Platform;
import taboolib.common.platform.PlatformSide;
import taboolib.common.platform.Plugin;

import java.io.File;

/**
 * TabooLib
 * taboolib.platform.SpongePlugin
 *
 * @author sky
 * @since 2021/6/26 8:39 下午
 */
@org.spongepowered.api.plugin.Plugin(
        id = "@plugin_id@",
        name = "@plugin_name@",
        version = "@plugin_version@"
)
@PlatformSide(Platform.SPONGE)
public class SpongePlugin {

    @Nullable
    private static final Plugin pluginInstance;
    private static SpongePlugin instance;

    @Inject
    private PluginContainer pluginContainer;

    @Inject
    @ConfigDir(sharedRoot = false)
    private File pluginConfigDir;

    static {
        TabooLibCommon.lifeCycle(LifeCycle.CONST);
        pluginInstance = IOKt.findInstance(Plugin.class);
    }

    public SpongePlugin() {
        instance = this;
        TabooLibCommon.lifeCycle(LifeCycle.INIT);
    }

    @Listener
    private void e(GamePreInitializationEvent e) {
        TabooLibCommon.lifeCycle(LifeCycle.LOAD);
        if (pluginInstance != null) {
            pluginInstance.onLoad();
        }
    }

    @Listener
    private void e(GameInitializationEvent e) {
        TabooLibCommon.lifeCycle(LifeCycle.ENABLE);
        if (pluginInstance != null) {
            pluginInstance.onEnable();
        }
        FunctionKt.startExecutor();
    }

    @Listener
    private void e(GameStartedServerEvent e) {
        TabooLibCommon.lifeCycle(LifeCycle.ACTIVE);
        if (pluginInstance != null) {
            pluginInstance.onActive();
        }
    }

    @Listener
    private void e(GameStoppedServerEvent e) {
        TabooLibCommon.lifeCycle(LifeCycle.DISABLE);
        if (pluginInstance != null) {
            pluginInstance.onDisable();
        }
    }

    public static SpongePlugin getInstance() {
        return instance;
    }

    public PluginContainer getPluginContainer() {
        return pluginContainer;
    }

    public File getPluginConfigDir() {
        return pluginConfigDir;
    }
}

File: platform-sponge/src/main/kotlin/taboolib/platform/SpongePlugin.kt
package taboolib.platform

import com.google.inject.Inject
import org.spongepowered.api.config.ConfigDir
import org.spongepowered.api.event.Listener
import org.spongepowered.api.event.game.GameReloadEvent
import org.spongepowered.api.event.game.state.GameConstructionEvent
import org.spongepowered.api.event.game.state.GameInitializationEvent
import org.spongepowered.api.event.game.state.GameStartedServerEvent
import org.spongepowered.api.event.game.state.GameStoppingServerEvent
import org.spongepowered.api.plugin.Plugin
import org.spongepowered.api.plugin.PluginContainer
import taboolib.common.TabooLibCommon
import taboolib.common.io.findInstance
import taboolib.common.platform.startExecutor
import java.io.File

/**
 * TabooLib
 * taboolib.platform.SpongePlugin
 *
 * @author sky
 * @since 2021/6/15 1:54 上午
 */
@Plugin(id = "@plugin_id@", name = "@plugin_name@", version = "@plugin_version@")
class SpongePlugin {

    val pluginInstance: taboolib.plugin.Plugin?

    @Inject
    lateinit var pluginContainer: PluginContainer

    @Inject
    @ConfigDir(sharedRoot = false)
    lateinit var pluginConfigDir: File

    init {
        TabooLibCommon.init()
        instance = this
        pluginInstance = findInstance(taboolib.plugin.Plugin::class.java)
    }

    @Listener
    fun e(e: GameConstructionEvent) {
        pluginInstance?.onLoad()
    }

    @Listener
    fun e(e: GameInitializationEvent) {
        pluginInstance?.onEnable()
        startExecutor()
    }

    @Listener
    fun e(e: GameStartedServerEvent) {
        pluginInstance?.onActive()
    }

    @Listener
    fun e(e: GameStoppingServerEvent) {
        pluginInstance?.onDisable()
        TabooLibCommon.cancel()
    }

    companion object {

        lateinit var instance: SpongePlugin
            private set
    }
}